src/user-pool.ts

```diff
@@ -0,0 +1,572 @@
+   *     return event;
+   *   })
+   * });
+   * ```
+   *
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-email-sender.html
+   * @default - no trigger configured
+   */
+  readonly customEmailSender?: Trigger<lambda.CustomEmailSenderTriggerEvent>;
+  /**
+   * Amazon Cognito invokes this trigger to send SMS notifications to users.
+   *
+   * ```ts
+   * new UserPool(scope, id, {
+   *   customSmsSender: new Function(scope, id, async (event) => {
+   *     // handle event, set response properties, etc.
+   *     event.response... = ..
+   *     return event;
+   *   })
+   * });
+   * ```
+   *
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-sms-sender.html
+   * @default - no trigger configured
+   */
+  readonly customSmsSender?: Trigger<lambda.CustomSMSSenderTriggerEvent>;
+}
+
+/**
+ * Map a friendly trigger name to the {@link aws_cognito.UserPoolOperation}.
+ */
+const triggers: {
+  [triggerName in keyof UserPoolTriggers]: aws_cognito.UserPoolOperation;
+} = {
+  createAuthChallenge: aws_cognito.UserPoolOperation.CREATE_AUTH_CHALLENGE,
+  customEmailSender: aws_cognito.UserPoolOperation.CUSTOM_EMAIL_SENDER,
+  customMessage: aws_cognito.UserPoolOperation.CUSTOM_MESSAGE,
+  customSmsSender: aws_cognito.UserPoolOperation.CUSTOM_SMS_SENDER,
+  defineAuthChallenge: aws_cognito.UserPoolOperation.DEFINE_AUTH_CHALLENGE,
+  postAuthentication: aws_cognito.UserPoolOperation.POST_AUTHENTICATION,
+  postConfirmation: aws_cognito.UserPoolOperation.POST_CONFIRMATION,
+  preAuthentication: aws_cognito.UserPoolOperation.PRE_AUTHENTICATION,
+  preSignUp: aws_cognito.UserPoolOperation.PRE_SIGN_UP,
+  preTokenGeneration: aws_cognito.UserPoolOperation.PRE_TOKEN_GENERATION,
+  userMigration: aws_cognito.UserPoolOperation.USER_MIGRATION,
+  verifyAuthChallengeResponse:
+    aws_cognito.UserPoolOperation.VERIFY_AUTH_CHALLENGE_RESPONSE,
+};
+
+export interface UserPoolProps
+  extends Omit<aws_cognito.UserPoolProps, "lambdaTriggers"> {
+  /**
+   * Lambda functions to use for supported Cognito triggers.
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-identity-pools-working-with-aws-lambda-triggers.html
+   * @default - No Lambda triggers.
+   */
+  readonly lambdaTriggers?: UserPoolTriggers;
+}
+
+/**
+ * Defines an AWS Cognito User Pool.
+ */
+export class UserPool {
+  public static readonly FunctionlessType = "UserPool";
+  public readonly FunctionlessType = UserPool.FunctionlessType;
+
+  /**
+   * Create a {@link UserPool} from a pre-existing {@link aws_cognito.UserPool}.
+   *
+   * @param resource a pre-existing CDK User Pool L2 Construct.
+   * @returns a Functionless {@link UserPool}.
+   */
+  public static from(resource: aws_cognito.UserPool): UserPool {
+    return new UserPool(resource);
+  }
+
+  /**
+   * The underlying {@link aws_cognito.UserPool} instance.
+   */
+  readonly resource: aws_cognito.UserPool;
```

Comment:
Was going to ask if this should be `IUserPool`, but I see that `addTrigger` is on `UserPool` only. Thats annoying. I bet imported user pools cannot have triggers. 

Do we have a case where we'd want a user pool to be wrapped even if imported? (I think so?)

Reply:
https://github.com/aws/aws-cdk/blob/v1.169.0/packages/@aws-cdk/aws-cognito/lib/user-pool.ts#L708

Reply:
```ts
functionless.UserPool.from(aws_cognito.UserPool.fromUserPoolArn(...))
```

or

```ts
functionless.UserPool.fromUserPoolArn(...)
```

Reply:
That said... step function relies on `StepFunction` not `IStepFunction` because `IStepFunction` doesn't support `stateMachineType` and thus cannot be validated... maybe we need to fix that.